### PR TITLE
Advanced Merging

### PR DESCRIPTION
**Pull request by AI-Casanova**
Early draft of Advanced Merging functions adapted from [sd_meh](https://github.com/s1dlx/meh)

Working:
- Basic Merging
- Preset Block Merging
- SD1.5 git re-basin

Untested:
- Manual Block Merging

To be added:
- Block Weight visualizer (line graph of alpha and beta settings)
- Front End verification of user inputs
-  SDXL git re-basin (needs a complete new permutation spec)

### CODE REVIEW

**Comment by vladmandic** [comment]
where are we with this pr?  
its still in draft and not really up to date with dev anymore  
which functions are missing/broken? cant test all, there are sooo many  

code comments:

- python is notoriously unsafe with modules name handling, can you rename code files to be more unique?  
  e.g. `model.py` -> `model_merge.py`, `presets.py` -> `presets_merge.py` and `utils.py` -> `utils_merge.py`
- `log_vram` is unsafe, can you replace with `memstats/memory_stats()`?
- `if self.model_path.suffix == ".safetensors":` this may be unsafe as PathLike may or may not have suffix, better use `os.path.splitext`  

but those are pretty much nitpicks, the code looks pretty good to me!

comments without looking at the code, simply clicking through the ui...

- just name it "merge" and disable previous legacy "merge" :)
- why have device selector? "Device cpu/cuda"  
  afaik, this only impacts model load location and its really cpu in 99% of cases  
- really need hints :)  
  define in `html/en_locale.json`  
  not for every option in dropdown, but for core options at least  
- perhaps even add some `gr.HTML` elements with comments?  
  take a look at updated `modules/ui_train.py` for example  
- you know which merge method i'd LOVE to see?  
  **model** as-is with **lora:weight** where it can be any number of loras  
  basically pre-baking loras into model  
- save metadata  
  should be simple to view metadata for a model (using `gr.JSON` element)  
  and then optionally edit it before saving it
- additionally we should add our own metadata to any models created with this  
- preset block merge has `SDXL` checkbox  
  cant we autodetect that based on primary model?
- alpha block weight preset lists all presets  
  regardless if sdxl is selected or not

**Comment by AI-Casanova** [comment]
Device does save some amount of time because of parallelization. 

- For a simple SD1.5 merge time goes from 0:06 to 0:00
- A more complex merge type like euclidean add difference is 0:30 to 0:01
- SDXL Models are too big for VRAM (I'm running a 15gb T4) but I'm going to add a shuffle to VRAM option that takes Euclidean add difference on SDXL from 1:20 to 0:50

**Comment by vladmandic** [comment]
re: device - thats a good enough of a reason for sure.

**Comment by AI-Casanova** [comment]
Made some decent strides today, but have to log off

re: SDXL is there a good way of determining that from the front end? I wasn't aware of one.

**Comment by vladmandic** [comment]
quick feedback:
- use `devices.gc.collect(force=True)` instead of `gc.collect()`
- maybe add checkbox to call `sd_models.unload_model_weights()` before merge?
- `merge_mode_help` button doesnt need to exist at all, `merge_mode` itself can update `merge_mode_docs` on input change
- `merge_mode_docs` should be `gr.HTML` control instead of textbox (just remove `interactive` from `gr.update`)

re: SDXL - not 100% reliable before model is loaded since there are no keys and embedded metadata is unreliable at best  
but `sd_models.detect_pipeline` does a decent job for a hack.

**Comment by AI-Casanova** [comment]
I was trying to keep the UI clean by only offering the help box when requested, but I can definitely change that functionality

re: `sd_models.detect_pipeline` I accidentally merged a fp32 SDXL model(11+ gb), and it was detected as SD1.5

Which brings up the point, is there any particular reason to offer fp32 outputs?

**Comment by vladmandic** [comment]
send me exact model size & i'll add it to known list.
whats the point of fp32 models? none anymore as far as i know, but i'm someone will claim that their model looks better in fp32.

**Comment by vladmandic** [comment]
just occured to me, why not device `cpu` and `devices.device`, not hard-code `cuda`. on cuda platforms, its the same thing, but this way it would at least have a chance on non-cuda platforms (not that we have to have all that supported day-1).

**Comment by AI-Casanova** [comment]
I'll dig into `devices.device` and see if I can get the backend to comply

**Comment by vladmandic** [comment]
Btw, `os.path.splitext`returns tuple so you need `[1]` for actual extension.

**Comment by vladmandic** [comment]
Quick one - can you prefix all logging with `Merge: ` and use error levels? 
Especially stuff like `shared.log.info(e), it should be `shared.log.warning(f'Merge: {e}')`

**Comment by AI-Casanova** [comment]
Getting close to pulling the draft label

We may want to squash the merge, because my commit history is a bit insane

**Comment by AI-Casanova** [comment]
Requesting a deeper review now:

Stretch Goals for future dev cycle:
- Automatic model naming (and overwrite safety)
- SDXL Autodetect
- Custom Metadata
- Block Weight visualizer (line graph of alpha and beta settings)
- Front End verification of user inputs
- SDXL git re-basin (needs a complete new permutation spec)

Let me know if any of those need to be pulled up to this release

**Comment by vladmandic** [comment]
if you agree, i'd suggest to merge to `dev` and continue working there?
that way i can take everything under (7) and you can focus on few remaining things (1-6)?  
and there is no need for another review anyhow, now its time to just iron out quirks and dev branch is a good place for that.

re: merge vs squash - your choice? this pr currently has 19 commits which is not that horrible - i'm fine either way.

re: strech goals - totally agree those are not a must for v1
if you want to keep working on them, it can be done over time  

from everything on the list, i'd put one that is missing at the top: option "load merged model" and making "save model" optional checkbox.
that way you can merge models in memory and use it to generate samples without saving it to disk.

other than that, i'd say that moving to "lora merge" would be more beneficial than continuing with all possible items here.

notes:

1. **Bake in VAE** if its is set to None, it will use VAE from primary model, right?  
   if that's the case, lets call it "Replace VAE". There should never be a case where we produce a model without VAE.
2. **Merge device**: it still refers to `cuda` - can we use `gpu` instead?
3. why is there forced `sd_models.unload_model_weights` if `cuda` and no unload at all otherwise?  
   it should be a checkbox that applies to all cases, right next to "Save metadata"  
   (you pick if you want it enabled or disabled as default)
4. getting a warning on very basic weighted sum merge:

        WARNING  Merge: 'alpha_base'

5. **Bake in VAE** produces an error:

        ❱ 137 | if theta_0_key in theta_0:
        NotImplementedError: TensorDict does not support membership checks with the `in` keyword. If you want to check if a particular key is in your TensorDict, please use `key in tensordict.keys()` instead.

6. after completing simple merge, i switched to presets merge, selected `smoothstep` (as an example) and clicked merge - it errored out, but worked as soon as i reselected primary/secondary model in pulldowns (to same models)

        ❱ 64 │ "model_a": sd_models.checkpoints_list[kwargs.get("primary_model_name", None)].
        KeyError: 'cyan-2.5d-v1'

    i think issue is if model is displayed with or without shorthash in title which can change on model refresh  
    best to replace `sd_models.checkpoints_list[name]` access with `sd_models.get_closet_checkpoint_match(name)`

7. log messages and progress-bar are not styled  
  (but i can take that afterwards so you don't have to deal with that if you'd like)
  - sdnext uses styled pbar wherever possible  
    number of steps is something like `1131` which is what?  
    and performance measurement is by default in `it/s` which makes no sense for merge (e.g. my log shows `1956995.80it/s`)  
  - pbar desc should mean something  
    what is "stage 1" vs "stage 2"? what is "un-prune model"?
  - only very few log messages are prefixed with `Merge:`, but most are not really readable  
    e.g., i dont know what "remove thetas VRAM" means?
  - minor typos like model name ending with `.`

        Merge saved: /mnt/f/Models/Stable-diffusion/test3.safetensors.

**Comment by AI-Casanova** [comment]
Fixing in dev works for me, merge when ready: (you choice on merge flavor)

1. Will confirm VAE takes primary, and rename option
2. "GPU" is better yes
3. Unload is only necessary if using full VRAM (shuffling only loads one key at a time) will add checkbox that only shows on that option
4. I'll clarify the error messages and update hints. It's trying to tell you that you have a missing value in the base key)
5. I need to pull the `.todict()` above bake in, simple fix
6. I was going to ask about that and it slipped my mind `sd_models.get_closet_checkpoint_match(name)` is perfect
7. `1131` is total number of merged keys. The slower methods need some sort of pbar but it could be a percentage bar instead, the key number isn't really necessary to expose. Stage 1 is the only necessary tqdm. Unprune refers to the fact that only non-ema and necessary keys are merged, and the remaining keys in model 1 are added back afterwards. (it gets tricky to merge when you have unbalanced key lists between models). Honestly I think we should probably dump excess keys, EMA isn't accurate once merged, and excess keys like `ema_update` add no value.

**Comment by vladmandic** [comment]
merged!
re: tqdm - yup, i know, i was just pointing what user would see. i'll redo those, no worries.

**Comment by AI-Casanova** [comment]
I did just swap the unprune tqdms for conditional `log.info` but we may need to adjust that because the primary model should always show now (I disabled VAE merge)
